### PR TITLE
chore(net): Add `proptest` roundtrip to `rlp`  types

### PR DESCRIPTION
**Pull request by joshieDo**
* Adds `arbitrary/proptest::arbitrary` traits to types that derive `compact` and `rlp` traits.
* Adds auto tests to both. If a type has both traits, then it tests both inside the same `proptest!` closure.
* Two types needed manual arbitrary impl: `Chain` and `Capability` since they depend on outside types.
* One type needed manual arbitrary impl: `TransactionSigned` because the hash field is recalculated on `decode`, which wont match the random hash field generated by arbitrary.
* reth-primitives:  change `ethers-core::{H512,  U128}`  to own implementation for `H512` and `ruint::U128`
* change a couple more `bytes::Bytes` to `reth-primitives::Bytes`
* better UX for `main_codec` and `derive_arbitray`. Left some comments with examples below.

### CODE REVIEW

File: crates/primitives/src/transaction/mod.rs [hunk 1]
```diff
@@ -525,7 +534,7 @@ impl Decodable for TransactionKind {
 }
 
 /// Signed transaction.
-#[main_codec]
+#[main_codec(no_arbitrary)]
```
Comment (joshieDo) [inline crates/primitives/src/transaction/mod.rs:537]: example

File: crates/primitives/src/transaction/access_list.rs [hunk 2]
```diff
@@ -15,6 +15,6 @@ pub struct AccessListItem {
 }
 
 /// AccessList as defined in EIP-2930
-#[main_codec]
+#[main_codec(rlp)]
```
Comment (joshieDo) [inline crates/primitives/src/transaction/access_list.rs:18]: example

File: crates/primitives/src/transaction/tx_type.rs [hunk 3]
```diff
@@ -1,8 +1,8 @@
-use reth_codecs::{derive_compact_arbitrary, Compact};
+use reth_codecs::{derive_arbitrary, Compact};
 use serde::{Deserialize, Serialize};
 
 /// Transaction Type
-#[derive_compact_arbitrary]
+#[derive_arbitrary(compact)]
```
Comment (joshieDo) [inline crates/primitives/src/transaction/tx_type.rs:5]: example

File: crates/net/eth-wire/src/capability.rs [hunk 4]
```diff
@@ -26,6 +33,7 @@ pub enum CapabilityMessage {
 }
 
 /// A message indicating a supported capability and capability version.
+#[add_arbitrary_tests(rlp)]
```
Comment (joshieDo) [inline crates/net/eth-wire/src/capability.rs:36]: example

File: crates/primitives/src/transaction/mod.rs [hunk 5]
```diff
@@ -525,7 +534,8 @@ impl Decodable for TransactionKind {
...
+#[add_arbitrary_tests(rlp, compact)]
```
Comment (joshieDo) [inline crates/primitives/src/transaction/mod.rs:538]: example

File: crates/storage/codecs/derive/src/arbitrary.rs [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+        tests = quote! {
+            #[allow(non_snake_case)]
+            #[cfg(test)]
+            mod #mod_tests {
+                #(#traits)*
+
+                #[test]
+                fn proptest() {
+                    proptest::proptest!(proptest::prelude::ProptestConfig::with_cases(1000), |(field: super::#type_ident)| {
```
Comment (joshieDo) [inline crates/storage/codecs/derive/src/arbitrary.rs:54]: What would be the default number of cases?

Fields with `Vec<Vec<_>>` usually take quite long... Would an extra `arg` to define custom number of cases be useful for these types?

File: crates/primitives/src/receipt.rs [hunk 7]
```diff
@@ -112,7 +112,7 @@ impl Encodable for Receipt {
         if matches!(self.tx_type, TxType::EIP1559 | TxType::EIP2930) {
             payload_len += 1;
             // we include a string header for typed receipts, so include the length here
-            payload_len = length_of_length(payload_len);
+            payload_len += length_of_length(payload_len);
```
Comment (mattsse) [inline crates/primitives/src/receipt.rs:115]: is this an unrelated fix?

File: crates/storage/codecs/derive/src/arbitrary.rs (hunk 6 above)
Comment (mattsse) [reply crates/storage/codecs/derive/src/arbitrary.rs:54]: could we add this as an arg to the proc-macro?

File: crates/primitives/src/receipt.rs (hunk 7 above)
Comment (Rjected) [reply crates/primitives/src/receipt.rs:115]: yeah this is #847

**Review by gakonst — APPROVED** [review]

File: crates/storage/codecs/derive/src/arbitrary.rs (hunk 6 above)
Comment (gakonst) [reply crates/storage/codecs/derive/src/arbitrary.rs:54]: I don't mind doing in follow up

**Comment by gakonst** [comment]
@joshieDo can merge pending conflict

File: crates/storage/codecs/derive/src/arbitrary.rs (hunk 6 above)
Comment (joshieDo) [reply crates/storage/codecs/derive/src/arbitrary.rs]: we have to do now because it really blows up. Some types will have many vec structs of vec structs...

**Comment by joshieDo** [comment]
So added an `arg` for proptest cases, and limited some to fewer cases - since it was really exploding from all the `vec^N`. There are 2 types that even with just one case it's taking a lot of time.

Also made the default number of cases to `256` (same as proptest). Open to increase it, since the problematic cases should already be adjusted.

File: crates/primitives/src/transaction/mod.rs [hunk 8]
```diff
@@ -538,6 +548,53 @@ pub struct TransactionSigned {
...
+impl proptest::arbitrary::Arbitrary for TransactionSigned {
+    type Parameters = ();
+    type Strategy = BoxedStrategy<TransactionSigned>;
+
+    fn arbitrary_with(_: Self::Parameters) -> Self::Strategy {
+        any::<(Transaction, Signature)>()
+            .prop_map(move |(mut transaction, sig)| {
+                if let Some(chain_id) = transaction.chain_id().cloned() {
+                    // Otherwise we might overflow when calculating `v` on `recalculate_hash`
+                    transaction.set_chain_id(chain_id % (u64::MAX / 2 - 36));
```
[the remark this answers is not in the document]
Comment (joshieDo) [reply crates/primitives/src/transaction/mod.rs:567]: probably, now i got reminded of https://research.kudelskisecurity.com/2020/07/28/the-definitive-guide-to-modulo-bias-and-how-to-avoid-it/ lol

**Review by gakonst — APPROVED** [review]
Feeling good about it. Let's merge and we can iterate if needed. I think this is enough work to feel convinced about our ser/de being proper across encodings.
